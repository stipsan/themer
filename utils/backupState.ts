// Handy util for backing up state to the current URL

export function backupState(key: string, value: string) {
  const url = new URL(window.location.href)
  url.searchParams.set(key, value)
  window.history.replaceState({}, '', url.href)
}
