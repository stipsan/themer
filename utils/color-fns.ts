// https://github.com/sanity-io/design/blob/804bf73dffb1c0ecb1c2e6758135784502768bfe/packages/%40sanity/ui/src/theme/studioTheme/helpers.ts#L3-L17

import {
  multiply as _multiply,
  parseColor,
  rgbToHex,
  screen as _screen,
} from '@sanity/ui'

export function multiply(bg: string, fg: string): string {
  const b = parseColor(bg)
  const s = parseColor(fg)
  const hex = rgbToHex(_multiply(b, s))

  return hex
}

export function screen(bg: string, fg: string): string {
  const b = parseColor(bg)
  const s = parseColor(fg)
  const hex = rgbToHex(_screen(b, s))

  return hex
}
