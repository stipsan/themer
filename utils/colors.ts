import { black, hues, white } from '@sanity/color'

export const lightest = white.hex.toLowerCase()
export const darkest = black.hex.toLowerCase()

export const NEUTRAL_TONES = ['default', 'transparent']

export const { blue, cyan, gray, green, magenta, orange, purple, red, yellow } =
  hues
