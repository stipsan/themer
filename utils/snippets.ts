// Automatically generated by running `npm run build:snippets`

export function snippet(id: 'import-dynamic-js'): (first: string) => string
export function snippet(id: 'import-dynamic-ts'): (first: string) => string
export function snippet(id: 'import-static'): (first: string) => string
export function snippet(id: 'studio-config'): (first: string) => string
export function snippet(id: 'sanity.cli.ts'): () => string
export function snippet(id: 'sanity.cli.js'): () => string
export function snippet(
  id: 'studio-config-create-theme'
): (first: string) => string
export function snippet(
  id: 'import-create-theme-static'
): (first: string) => string
export function snippet(
  id: 'import-create-theme-dynamic'
): (first: string) => string
export function snippet(id: 'themer.d.ts'): (first: string) => string
export function snippet(id: 'tsconfig'): () => string
export function snippet(id: '_document.tsx'): (first: string) => string
export function snippet(id: '_document.js'): (first: string) => string
export function snippet(id) {
  switch (id) {
    case 'import-dynamic-js':
      return (first: string) => `const {theme} = await import(${first})`

    case 'import-dynamic-ts':
      return (first: string) => `const {theme} = (await import(
  // @ts-expect-error -- TODO setup themer.d.ts to get correct typings
  ${first}
)) as {theme: import('sanity').StudioTheme}`

    case 'import-static':
      return (first: string) => `// Add this URL ESM import
import {theme} from ${first}`

    case 'studio-config':
      return (first: string) => `// Add two lines of code to your workspace
import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'

import {schemaTypes} from './schemas'

// 1. Add the import
${first}

export default createConfig({
  theme, // <-- 2. add the theme here

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})`

    case 'sanity.cli.ts':
      return () => `// Change target to allow top-level await in sanity.config.ts
import {createCliConfig} from 'sanity/cli'
import type {UserConfig} from 'vite'

export default createCliConfig({
  api: {projectId: 'b5vzhxkv', dataset: 'production'},
  vite: (config: UserConfig): UserConfig => ({
    ...config,
    build: {...config.build, target: 'esnext'}
  })
})`

    case 'sanity.cli.js':
      return () => `// Change target to allow top-level await in sanity.config.js
import {createCliConfig} from 'sanity/cli'

export default createCliConfig({
  api: {projectId: 'b5vzhxkv', dataset: 'production'},
  vite: config => ({...config, build: {...config.build, target: 'esnext'}})
})`

    case 'studio-config-create-theme':
      return (
        first: string
      ) => `// Import createTheme and hues to quickly modify your theme without changing the import URL
import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'

import {schemaTypes} from './schemas'

${first}

export default createConfig({
  theme: createTheme({...hues, primary: {...hues.primary, mid: '#22fca8'}}),

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})`

    case 'import-create-theme-static':
      return (first: string) => `import {createTheme, hues} from ${first}`

    case 'import-create-theme-dynamic':
      return (first: string) => `const {createTheme, hues} = await import(
  ${first}
)`

    case 'themer.d.ts':
      return (first: string) => `module ${first} {
  interface Hue extends Omit<import('@sanity/color').ColorHueConfig, 'title' | 'midPoint'> {
    midPoint: 50 | 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900 | 950
  }
  interface Hues {
    default: Hue
    transparent: Hue
    primary: Hue
    positive: Hue
    caution: Hue
    critical: Hue
  }
  const hues: Hues
  type Theme = import('sanity').StudioTheme
  const createTheme = (hues: Hues): Theme => theme
  const theme: Theme

  export {hues, createTheme, theme}
}`

    case 'tsconfig':
      return () => `{
  "compilerOptions": {
    // target needs to be es2017 or newer to allow top-level await
    "target": "es2017",

    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true
  },
  "include": ["**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules"]
}`

    case '_document.tsx':
      return (
        first: string
      ) => `// As Studio v3 is in developer preview there's not yet a simple way to just add a <link> tag to the <head>
// Thus we have to re-implement DefaultDocument to make it happen.
// Expect this to get much easier before v3 hits stable

import React from 'react'
import {type DefaultDocumentProps} from 'sanity'
import {GlobalErrorHandler} from 'sanity/_unstable'

const globalStyles = ${'`'}
  html {
    background-color: #f1f3f6;
  }
  html,
  body,
  #sanity {
    height: 100%;
  }
  body {
    margin: 0;
    -webkit-font-smoothing: antialiased;
  }
${'`'}

interface FaviconProps {
  basePath: string
}

function Favicons({basePath}: FaviconProps) {
  const base = ${'`'}${'${'}basePath.replace(${/\/+$/}, '')}/static${'`'}
  return (
    <>
      <link rel="icon" href={${'`'}${'${'}base}/favicon.ico${'`'}} sizes="any" />
      <link rel="icon" href={${'`'}${'${'}base}/favicon.svg${'`'}} type="image/svg+xml" />
      <link rel="apple-touch-icon" href={${'`'}${'${'}base}/apple-touch-icon.png${'`'}} />
      <link rel="manifest" href={${'`'}${'${'}base}/manifest.webmanifest${'`'}} />
    </>
  )
}

const EMPTY_ARRAY: never[] = []
export default function DefaultDocument(props: DefaultDocumentProps): React.ReactElement {
  const {entryPath, css = EMPTY_ARRAY, basePath = '/'} = props
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover" />
        <meta name="robots" content="noindex" />
        <meta name="referrer" content="same-origin" />
        {/* This is the only line of code we're adding that is different from the default implementation of DefaultDocument */}
        <link
          rel="modulepreload"
          href={${first}}
        />

        <Favicons basePath={basePath} />

        <title>Sanity Studio</title>

        <GlobalErrorHandler />

        {css.map(href => (
          <link key={href} rel="stylesheet" href={href} />
        ))}
        <style>{globalStyles}</style>
      </head>
      <body>
        <div id="sanity" />
        <script type="module" src={entryPath} />
      </body>
    </html>
  )
}`

    case '_document.js':
      return (
        first: string
      ) => `// As Studio v3 is in developer preview there's not yet a simple way to just add a <link> tag to the <head>
// Thus we have to re-implement DefaultDocument to make it happen.
// Expect this to get much easier before v3 hits stable

import React from 'react'
import {GlobalErrorHandler} from 'sanity/_unstable'

const globalStyles = ${'`'}
  html {
    background-color: #f1f3f6;
  }
  html,
  body,
  #sanity {
    height: 100%;
  }
  body {
    margin: 0;
    -webkit-font-smoothing: antialiased;
  }
${'`'}

function Favicons({basePath}) {
  const base = ${'`'}${'${'}basePath.replace(${/\/+$/}, '')}/static${'`'}
  return (
    <>
      <link rel="icon" href={${'`'}${'${'}base}/favicon.ico${'`'}} sizes="any" />
      <link rel="icon" href={${'`'}${'${'}base}/favicon.svg${'`'}} type="image/svg+xml" />
      <link rel="apple-touch-icon" href={${'`'}${'${'}base}/apple-touch-icon.png${'`'}} />
      <link rel="manifest" href={${'`'}${'${'}base}/manifest.webmanifest${'`'}} />
    </>
  )
}

const EMPTY_ARRAY = []
export default function DefaultDocument(props) {
  const {entryPath, css = EMPTY_ARRAY, basePath = '/'} = props
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover" />
        <meta name="robots" content="noindex" />
        <meta name="referrer" content="same-origin" />
        {/* This is the only line of code we're adding that is different from the default implementation of DefaultDocument */}
        <link
          rel="modulepreload"
          href={${first}}
        />

        <Favicons basePath={basePath} />

        <title>Sanity Studio</title>

        <GlobalErrorHandler />

        {css.map(href => (
          <link key={href} rel="stylesheet" href={href} />
        ))}
        <style>{globalStyles}</style>
      </head>
      <body>
        <div id="sanity" />
        <script type="module" src={entryPath} />
      </body>
    </html>
  )
}`

    default:
      throw new TypeError('Unknown snippet id: ' + id)
  }
}
