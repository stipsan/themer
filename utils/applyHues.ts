/**
 * Yes a lot of the code here is gnarly, nasty and overly defensive.
 * That's because this file is currently in Stage 1 "Make it work".
 * The next two stages will be "Make right" and "Make it fast".
 * Stage 2 requires integrtation tests.
 */

import type { PartialDeep } from 'type-fest'
import { darkest, lightest } from 'utils/colors'
import type { Hues } from 'utils/types'

const defaultHues: Hues = {
  default: {
    lightest,
    darkest,
    mid: '#8690a0',
    midPoint: 500,
  },
  primary: {
    lightest,
    darkest,
    mid: '#2276fc',
    midPoint: 500,
  },
  transparent: {
    lightest,
    darkest,
    mid: '#8690a0',
    midPoint: 500,
  },
  positive: {
    lightest,
    darkest,
    mid: '#43d675',
    midPoint: 400,
  },
  caution: {
    lightest,
    darkest,
    mid: '#fbd024',
    midPoint: 300,
  },
  critical: {
    lightest,
    darkest,
    mid: '#f03e2f',
    midPoint: 500,
  },
}

export function applyHues(
  _hues: PartialDeep<Hues>,
  defaults: Hues = defaultHues
): Hues {
  // Filter out undefined keys etc
  const hues = JSON.parse(JSON.stringify(_hues))
  const defaultMid = hues.default?.mid?.toLowerCase() || defaults.default.mid
  const primaryMid = hues.primary?.mid?.toLowerCase() || defaults.primary.mid
  const transparentMid =
    hues.transparent?.mid?.toLowerCase() || defaults.transparent.mid
  const positiveMid = hues.positive?.mid?.toLowerCase() || defaults.positive.mid
  const cautionMid = hues.caution?.mid?.toLowerCase() || defaults.caution.mid
  const criticalMid = hues.critical?.mid?.toLowerCase() || defaults.critical.mid
  return {
    default: {
      ...defaults.default,
      midPoint:
        defaultMid === defaults.default.mid ? defaults.default.midPoint : 500,
      ...hues.default,
      mid: defaultMid,
    },
    primary: {
      ...defaults.primary,
      midPoint:
        primaryMid === defaults.primary.mid ? defaults.primary.midPoint : 500,
      ...hues.primary,
      mid: primaryMid,
    },
    transparent: {
      ...defaults.transparent,
      midPoint:
        transparentMid === defaults.transparent.mid
          ? defaults.transparent.midPoint
          : 500,
      ...hues.transparent,
      mid: transparentMid,
    },
    positive: {
      ...defaults.positive,
      midPoint:
        positiveMid === defaults.positive.mid
          ? defaults.positive.midPoint
          : 500,
      ...hues.positive,
      mid: positiveMid,
    },
    caution: {
      ...defaults.caution,
      midPoint:
        cautionMid === defaults.caution.mid ? defaults.caution.midPoint : 500,
      ...hues.caution,
      mid: cautionMid,
    },
    critical: {
      ...defaults.critical,
      midPoint:
        criticalMid === defaults.critical.mid
          ? defaults.critical.midPoint
          : 500,
      ...hues.critical,
      mid: criticalMid,
    },
  }
}
