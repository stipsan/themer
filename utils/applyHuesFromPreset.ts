import { applyHues } from 'utils/applyHues'
import { parseHuesFromSearchParams } from 'utils/parseHuesFromSearchParams'
import type { Hues } from 'utils/types'

export function applyHuesFromPreset(
  presetParams: URLSearchParams,
  searchParams: URLSearchParams
): Hues {
  const presetHues = applyHues(parseHuesFromSearchParams(presetParams))

  return applyHues(parseHuesFromSearchParams(searchParams), presetHues)
}
