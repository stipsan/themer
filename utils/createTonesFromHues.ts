import { type ColorTints, COLOR_TINTS } from '@sanity/color'
import { getColorHex } from 'utils/getColorHex'
import type { Hue, Hues } from 'utils/types'

// https://github.com/sanity-io/design/blob/804bf73dffb1c0ecb1c2e6758135784502768bfe/packages/%40sanity/color/scripts/generate.ts#L42-L58
function createTintsFromHue(hue: Hue, title: string): ColorTints {
  const initial = {} as ColorTints
  const tints = COLOR_TINTS.reduce((acc, tint) => {
    acc[tint] = {
      title: `${title} ${tint}`,
      hex: getColorHex(hue, tint),
    }

    return acc
  }, initial)

  return tints
}

export function createTonesFromHues(hues: Hues): {
  default: ColorTints
  primary: ColorTints
  transparent: ColorTints
  positive: ColorTints
  caution: ColorTints
  critical: ColorTints
} {
  return {
    default: createTintsFromHue(hues.default, 'Default'),
    primary: createTintsFromHue(hues.primary, 'Primary'),
    transparent: createTintsFromHue(hues.transparent, 'Transparent'),
    positive: createTintsFromHue(hues.positive, 'Positive'),
    caution: createTintsFromHue(hues.caution, 'Caution'),
    critical: createTintsFromHue(hues.critical, 'Critical'),
  }
}
