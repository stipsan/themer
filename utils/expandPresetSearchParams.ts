import { TONES } from 'utils/colors'
import { roundMidPoint } from 'utils/roundMidPoint'
import { stringifyColorSearchParam } from 'utils/stringifyColorSearchParam'
import type { Hues } from 'utils/types'

export function expandPresetSearchParams(
  searchParams: URLSearchParams,
  hues: Hues
) {
  for (const tone of TONES) {
    const hue = hues[tone]
    searchParams.set(
      tone,
      `${stringifyColorSearchParam(hue.mid)};${roundMidPoint(
        hue.midPoint
      )};lightest:${stringifyColorSearchParam(
        hue.lightest
      )};darkest:${stringifyColorSearchParam(hue.darkest)}`
    )
  }
}
