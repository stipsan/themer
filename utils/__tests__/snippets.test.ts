import JSON5 from 'json5'
import { snippet, snippets } from 'utils/snippets'

for (const id of snippets) {
  test(id, () => {
    expect(
      (snippet(id as any) as any)(JSON5.stringify('${first}}'))
    ).toMatchSnapshot()
  })
}
