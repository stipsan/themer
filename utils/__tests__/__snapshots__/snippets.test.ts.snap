// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_document.js 1`] = `
"// This is to generate a <link rel=\\"modulepreload\\" href='\${first}}'> to the <head>
// As Studio v3 is in developer preview there's not yet a simple way to just add a <link> tag to the <head>
// Thus we have to re-implement DefaultDocument to make it happen.
// Expect this to get much easier before v3 hits stable

import React from 'react'
import {GlobalErrorHandler} from 'sanity/_unstable'

const globalStyles = \`
  html {
    background-color: #f1f3f6;
  }
  html,
  body,
  #sanity {
    height: 100%;
  }
  body {
    margin: 0;
    -webkit-font-smoothing: antialiased;
  }
\`

function Favicons({basePath}) {
  const base = \`\${basePath.replace(/\\\\/+$/, '')}/static\`
  return (
    <>
      <link rel=\\"icon\\" href={\`\${base}/favicon.ico\`} sizes=\\"any\\" />
      <link rel=\\"icon\\" href={\`\${base}/favicon.svg\`} type=\\"image/svg+xml\\" />
      <link rel=\\"apple-touch-icon\\" href={\`\${base}/apple-touch-icon.png\`} />
      <link rel=\\"manifest\\" href={\`\${base}/manifest.webmanifest\`} />
    </>
  )
}

const EMPTY_ARRAY = []
export default function DefaultDocument(props) {
  const {entryPath, css = EMPTY_ARRAY, basePath = '/'} = props
  return (
    <html lang=\\"en\\">
      <head>
        <meta charSet=\\"utf-8\\" />
        <meta
          name=\\"viewport\\"
          content=\\"width=device-width, initial-scale=1, viewport-fit=cover\\"
        />
        <meta name=\\"robots\\" content=\\"noindex\\" />
        <meta name=\\"referrer\\" content=\\"same-origin\\" />
        {/* This is the only line of code we're adding that is different from the default implementation of DefaultDocument */}
        <link
          rel=\\"modulepreload\\"
          href={'\${first}}'}
        />

        <Favicons basePath={basePath} />

        <title>Sanity Studio</title>

        <GlobalErrorHandler />

        {css.map(href => (
          <link key={href} rel=\\"stylesheet\\" href={href} />
        ))}
        <style>{globalStyles}</style>
      </head>
      <body>
        <div id=\\"sanity\\" />
        <script type=\\"module\\" src={entryPath} />
      </body>
    </html>
  )
}"
`;

exports[`_document.tsx 1`] = `
"// This is to generate a <link rel=\\"modulepreload\\" href='\${first}}'> to the <head>
// As Studio v3 is in developer preview there's not yet a simple way to just add a <link> tag to the <head>
// Thus we have to re-implement DefaultDocument to make it happen.
// Expect this to get much easier before v3 hits stable

import React from 'react'
import {type DefaultDocumentProps} from 'sanity'
import {GlobalErrorHandler} from 'sanity/_unstable'

const globalStyles = \`
  html {
    background-color: #f1f3f6;
  }
  html,
  body,
  #sanity {
    height: 100%;
  }
  body {
    margin: 0;
    -webkit-font-smoothing: antialiased;
  }
\`

interface FaviconProps {
  basePath: string
}

function Favicons({basePath}: FaviconProps) {
  const base = \`\${basePath.replace(/\\\\/+$/, '')}/static\`
  return (
    <>
      <link rel=\\"icon\\" href={\`\${base}/favicon.ico\`} sizes=\\"any\\" />
      <link rel=\\"icon\\" href={\`\${base}/favicon.svg\`} type=\\"image/svg+xml\\" />
      <link rel=\\"apple-touch-icon\\" href={\`\${base}/apple-touch-icon.png\`} />
      <link rel=\\"manifest\\" href={\`\${base}/manifest.webmanifest\`} />
    </>
  )
}

const EMPTY_ARRAY: never[] = []
export default function DefaultDocument(
  props: DefaultDocumentProps
): React.ReactElement {
  const {entryPath, css = EMPTY_ARRAY, basePath = '/'} = props
  return (
    <html lang=\\"en\\">
      <head>
        <meta charSet=\\"utf-8\\" />
        <meta
          name=\\"viewport\\"
          content=\\"width=device-width, initial-scale=1, viewport-fit=cover\\"
        />
        <meta name=\\"robots\\" content=\\"noindex\\" />
        <meta name=\\"referrer\\" content=\\"same-origin\\" />
        {/* This is the only line of code we're adding that is different from the default implementation of DefaultDocument */}
        <link
          rel=\\"modulepreload\\"
          href={'\${first}}'}
        />

        <Favicons basePath={basePath} />

        <title>Sanity Studio</title>

        <GlobalErrorHandler />

        {css.map(href => (
          <link key={href} rel=\\"stylesheet\\" href={href} />
        ))}
        <style>{globalStyles}</style>
      </head>
      <body>
        <div id=\\"sanity\\" />
        <script type=\\"module\\" src={entryPath} />
      </body>
    </html>
  )
}"
`;

exports[`import-create-theme-dynamic 1`] = `
"const {createTheme, hues} = await import(
  '\${first}}'
)"
`;

exports[`import-create-theme-static 1`] = `"import {createTheme, hues} from '\${first}}'"`;

exports[`import-dynamic-js 1`] = `"const {theme} = await import('\${first}}')"`;

exports[`import-dynamic-ts 1`] = `
"const {theme} = (await import(
  // @ts-expect-error -- TODO setup themer.d.ts to get correct typings
  '\${first}}'
)) as {theme: import('sanity').StudioTheme}"
`;

exports[`import-static 1`] = `"import {theme} from '\${first}}'"`;

exports[`next-config-build-time-js 1`] = `
"module.exports = {
  experimental: {urlImports: ['\${first}}']}
}"
`;

exports[`next-config-build-time-ts 1`] = `
"// @ts-check

/**
 * @type {import('next').NextConfig}
 **/
const nextConfig = {
  experimental: {urlImports: ['\${first}}']}
}

module.exports = nextConfig"
`;

exports[`pages/_document.js 1`] = `
"// This is necessary for SSR to work correctly and prevents broken CSS

import Document from 'next/document'
import {ServerStyleSheet} from 'styled-components'

export default class CustomDocument extends Document {
  static async getInitialProps(ctx) {
    const sheet = new ServerStyleSheet()
    const originalRenderPage = ctx.renderPage

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: App => props => sheet.collectStyles(<App {...props} />)
        })

      const initialProps = await Document.getInitialProps(ctx)
      return {
        ...initialProps,
        styles: [initialProps.styles, sheet.getStyleElement()]
      }
    } finally {
      sheet.seal()
    }
  }
}"
`;

exports[`pages/_document.tsx 1`] = `
"// This is necessary for SSR to work correctly and prevents broken CSS

import Document, {type DocumentContext} from 'next/document'
import {ServerStyleSheet} from 'styled-components'

export default class CustomDocument extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    const sheet = new ServerStyleSheet()
    const originalRenderPage = ctx.renderPage

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: App => props => sheet.collectStyles(<App {...props} />)
        })

      const initialProps = await Document.getInitialProps(ctx)
      return {
        ...initialProps,
        styles: [initialProps.styles, sheet.getStyleElement()]
      }
    } finally {
      sheet.seal()
    }
  }
}"
`;

exports[`pages-index 1`] = `
"// Loading the custom theme on the page level instead of in sanity.config

import Head from 'next/head'
import {useEffect, useState} from 'react'
import {Studio} from 'sanity'

import _config from '../sanity.config'

export default function IndexPage() {
  const [config, setConfig] = useState(_config)

  useEffect(
    // Start fetching the theme in parallel with the Studio auth loading
    () =>
      // The webpackIgnore tells webpack to not attempt bundling this dynamic import,
      // and instead let it run natively in the browser at runtime
      void import(
        /* webpackIgnore: true */ '\${first}}'
      ).then(({theme}) => setConfig(config => ({...config, theme}))),
    []
  )

  return <Studio config={config} />
}"
`;

exports[`sanity.cli.js 1`] = `
"// Change target to allow top-level await in sanity.config.js

import {createCliConfig} from 'sanity/cli'

export default createCliConfig({
  api: {projectId: 'b5vzhxkv', dataset: 'production'},
  vite: config => ({...config, build: {...config.build, target: 'esnext'}})
})"
`;

exports[`sanity.cli.ts 1`] = `
"// Change target to allow top-level await in sanity.config.ts

import {createCliConfig} from 'sanity/cli'
import type {UserConfig} from 'vite'

export default createCliConfig({
  api: {projectId: 'b5vzhxkv', dataset: 'production'},
  vite: (config: UserConfig): UserConfig => ({
    ...config,
    build: {...config.build, target: 'esnext'}
  })
})"
`;

exports[`studio-config 1`] = `
"// Add two lines of code to your workspace

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'

import {schemaTypes} from './schemas'

// 1. Add the import
'\${first}}'

export default createConfig({
  theme, // <-- 2. add the theme here

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-create-theme 1`] = `
"// Import createTheme and hues to quickly modify your theme without changing the import URL

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'

import {schemaTypes} from './schemas'

'\${first}}'

export default createConfig({
  theme: createTheme({...hues, primary: {...hues.primary, mid: '#22fca8'}}),

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-create-theme-static-import 1`] = `
"// Import createTheme and hues to quickly modify your theme without changing the import URL

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'

'\${first}}'

import {schemaTypes} from './schemas'

export default createConfig({
  theme: createTheme({...hues, primary: {...hues.primary, mid: '#22fca8'}}),

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-local-import 1`] = `
"// Add two lines of code to your workspace

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'
// 1. Add the impoort to the theme.js you downloaded
import {theme} from './theme'

import {schemaTypes} from './schemas'

export default createConfig({
  theme, // <-- 2. add the theme here

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-local-import-ts 1`] = `
"// Add the theme import and its typings to your workspace

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'

import {schemaTypes} from './schemas'

// 1. Add the impoort to the theme.js you downloaded
import {theme as _theme} from './theme'

// 2. Assign typings to the theme
const theme = _theme as import('sanity').StudioTheme

export default createConfig({
  theme, // <-- 3. add the theme here

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-next-runtime-1 1`] = `
"// There's no theme import in this file since we're handling that in a useEffect in the index page

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'
import {schemaTypes} from './schemas'

export default createConfig({
  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-next-runtime-2 1`] = `
"// Allow reading the default theme variables while the custom theme is loading

import {createConfig, defaultTheme} from 'sanity'
import {deskTool} from 'sanity/desk'
import {schemaTypes} from './schemas'

export default createConfig({
  theme: defaultTheme,

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`studio-config-static-import 1`] = `
"// Add two lines of code to your workspace

import {createConfig} from 'sanity'
import {deskTool} from 'sanity/desk'
// 1. Add the import
'\${first}}'

import {schemaTypes} from './schemas'

export default createConfig({
  theme, // <-- 2. add the theme here

  name: 'default',
  title: 'My Sanity Project',
  projectId: 'b5vzhxkv',
  dataset: 'production',
  plugins: [deskTool()],
  schema: {types: schemaTypes}
})"
`;

exports[`themer.d.ts 1`] = `
"module '\${first}}' {
  interface Hue
    extends Omit<import('@sanity/color').ColorHueConfig, 'title' | 'midPoint'> {
    midPoint: 50 | 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900 | 950
  }
  interface Hues {
    default: Hue
    transparent: Hue
    primary: Hue
    positive: Hue
    caution: Hue
    critical: Hue
  }
  const hues: Hues
  type Theme = import('sanity').StudioTheme
  const createTheme = (hues: Hues): Theme => theme
  const theme: Theme

  export {hues, createTheme, theme}
}"
`;

exports[`tsconfig 1`] = `
"{
  \\"compilerOptions\\": {
    // target needs to be es2017 or newer to allow top-level await
    \\"target\\": \\"es2017\\",

    \\"lib\\": [\\"dom\\", \\"dom.iterable\\", \\"esnext\\"],
    \\"allowJs\\": true,
    \\"skipLibCheck\\": true,
    \\"strict\\": true,
    \\"forceConsistentCasingInFileNames\\": true,
    \\"noEmit\\": true,
    \\"esModuleInterop\\": true,
    \\"module\\": \\"esnext\\",
    \\"moduleResolution\\": \\"node\\",
    \\"resolveJsonModule\\": true,
    \\"isolatedModules\\": true,
    \\"jsx\\": \\"preserve\\",
    \\"incremental\\": true
  },
  \\"include\\": [\\"**/*.ts\\", \\"**/*.tsx\\"],
  \\"exclude\\": [\\"node_modules\\"]
}"
`;
